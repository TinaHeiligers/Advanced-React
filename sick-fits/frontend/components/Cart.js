import React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import { adopt } from "react-adopt";
import User from "./User";
import CartStyles from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import SickButton from "./styles/SickButton";
import CartItem from "./CartItem";
import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";
// using the @client directive tells apollo that this is local and we're not going to the server for this
const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;
// we pass the render function in as a child to get rid of the proptypes error.
const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});
const Cart = props => (
  <Composed>
    {({ user, toggleCart, localState }) => {
      const me = user.data.me;
      if (!me) return null;
      return (
        <CartStyles open={localState.data.cartOpen}>
          <header>
            <CloseButton onClick={toggleCart} title="close">
              &times;
            </CloseButton>
            <Supreme>
              {`${me.name}${me.name.slice(-1) === "s" ? "'" : "'s"}`} Cart
            </Supreme>
            <p>
              You have {me.cart.length} Item
              {me.cart.length === 1 ? "" : "s"} in your cart.
            </p>
          </header>
          <ul>
            {me.cart.map(cartItem => (
              <CartItem key={cartItem.id} cartItem={cartItem}>
                {cartItem.id}
              </CartItem>
            ))}
          </ul>
          <footer>
            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
            <SickButton>Checkout</SickButton>
          </footer>
        </CartStyles>
      );
    }}
  </Composed>
);

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };
