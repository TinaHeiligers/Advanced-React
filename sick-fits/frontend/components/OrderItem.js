import React from "react";
import formatMoney from "../lib/formatMoney";

const OrderItem = props => {
  const item = props.item;
  return (
    <div className="order-item" key={item.id}>
      <img src={item.image} alt={item.title} />
      <div className="item-details">
        <h2>{item.title}</h2>
        <p>Qty: {item.quantity}</p>
        <p>Each: {formatMoney(item.price)}</p>
        <p>Subtotal: {formatMoney(item.price * item.quantity)}</p>
        <p>{item.description}</p>
      </div>
    </div>
  );
};

export default OrderItem;
