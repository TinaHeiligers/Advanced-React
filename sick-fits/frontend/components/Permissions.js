import { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types'
import Error from './ErrorMessage';
import Table from './styles/Table'
import SickButton from './styles/SickButton'

const possiblePermisssions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE',
];

const ALL_USERS_QUERY = gql`
  query {
    users {
      id
      name
      email
      permissions
    }
  }
`;

const Permissions = props => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) =>
    console.log(data) || (
      <div>
        <Error error={error} />
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermisssions.map((permission, idx) => (<th key={idx}>{permission}</th>))}
                <th>ðŸ‘‡</th>
              </tr>
            </thead>
            <tbody>{data.users.map((user, id) =>
              <UserPermissions key={id} user={user}/>
            )}</tbody>
          </Table>
        </div>
        <p>Hey!</p>
      </div>
    )}
  </Query>
);

class UserPermissions extends Component {
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string,
      id: PropTypes.string,
      permissions: PropTypes.array,
    }).isRequired,
  };
  // we're only using props to seed the initial state, hence not strictly violating the single source of truth pattern
  state = {
    permissions: this.props.user.permissions,
  }
  handlePermissionChange = (e) => {
    // either add or remove the permission to the array on state
    const checkbox = e.target;
    let updatedPermissions = [...this.state.permissions];
    // add or remove the permissions
    if (checkbox.checked) {
      updatedPermissions.push(checkbox.value);
    } else {
      updatedPermissions = updatedPermissions.filter((permission) => permission !== checkbox.value);
    }
    this.setState({ permissions: updatedPermissions });
  }
  render() {
    const user = this.props.user;
    return (
      <tr>
        <td>{user.name}</td>
        <td>{user.email}</td>
        { possiblePermisssions.map((permission, idx) =>
          <td key={idx}>
            <label htmlFor={`${user.id}-permission-${permission}`}>
              <input
                id={`${user.id}-permission-${permission}`}
                type="checkbox"
                checked={this.state.permissions.includes(permission)}
                value={permission}
                onChange={this.handlePermissionChange}/>
            </label>
          </td>
        )}
        <td>
          <SickButton>Update!</SickButton>
        </td>
      </tr>)
  }
}
export default Permissions;

