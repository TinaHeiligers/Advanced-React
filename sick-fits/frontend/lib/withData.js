import withApollo from 'next-with-apollo'; // gives a HOC that exposes the Apollo client to the child components
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
  });
}

export default withApollo(createClient);
// Request kinda like middleware to handle cookies
