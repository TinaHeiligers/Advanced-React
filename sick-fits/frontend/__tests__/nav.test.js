import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeUser } from "../lib/testUtils";
// create mocks outside of the tests
export const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } }
  }
];

export const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } }
  }
];

describe("<Nav />", () => {
  it("renders out a minimal nav when a user is not signed in", async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    // we'll use a snapshot to test this.
    const nav = wrapper.find('[data-test="nav"]');
    expect(toJSON(nav)).toMatchSnapshot();
    // console.log(wrapper.debug());
  });
});
