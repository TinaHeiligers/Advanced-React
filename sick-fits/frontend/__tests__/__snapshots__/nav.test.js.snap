// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders out a minimal nav when a user is not signed in 1`] = `
Array [
  <NavStyles
    data-test="nav"
  >
    <ul
      className="NavStyles-sc-11c0d2g-0 bjCNSW"
      data-test="nav"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
        >
          Shop
        </a>
      </Link>
      <Link
        href="/signup"
      >
        <a
          href="/signup"
          onClick={[Function]}
        >
          Sign In
        </a>
      </Link>
    </ul>
  </NavStyles>,
  <ul
    className="NavStyles-sc-11c0d2g-0 bjCNSW"
    data-test="nav"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
      >
        Shop
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        Sign In
      </a>
    </Link>
  </ul>,
]
`;
